### PR TITLE
Fast Delta Profiling

### PR DESCRIPTION
**Pull request by pmbauer**
## Description

The [pprof protobuf structure](https://github.com/google/pprof/blob/main/proto/profile.proto) when hydrated in-memory as golang structs has significant overhead in memory and allocations.  The structure has multiple nested self-referencing sub-structs that cause GC overhead to the point of dominating CPU resources for some services.

We need to keep this structure in memory for computing delta profiles between profile cycles, making our profiling client a cause for performance regressions in some cases.

This pull adds a new delta profile implementation - `fastdelta` - that uses a streaming pprof parser and low heap allocations so the entire hydrated pprof structure no longer needs to be kept in memory between profile cycles.

## Testing Done

- tested in reliability env with `DD_PROFILING_DELTA_METHOD=comparing`, no fidelity failures
- tested on production load with `DD_PROFILING_DELTA_METHOD=fastdelta`
- similar results to the synthetic benchmark (below) in tested production service and in reliability env

## Deploy Plan

1. one `dd-trace-go` release with both delta profile implementations, old algorithm as default, new algorithm enabled with `DD_PROFILING_DELTA_METHOD` env var. This is so we can continue testing internally.
2. release N+1, make `fastdelta` the default, enable old algorithm with an env var fallback if there are any issues.
3. at some point, we may remove the old algorithm entirely.

## Results

On synthetic benchmarks

- ~4x improvement in wall time
- ~10x improvement in memory overhead
- ~20x improvement in allocations

```
name                                  old time/op    new time/op    delta
FastDelta/testdata/heap.pprof-10        2.09ms ± 1%    0.52ms ± 0%  -75.00%  (p=0.008 n=5+5)
FastDelta/testdata/big-heap.pprof-10    31.6ms ± 1%     7.0ms ± 0%  -77.93%  (p=0.008 n=5+5)

name                                  old alloc/op   new alloc/op   delta
FastDelta/testdata/heap.pprof-10        2.98MB ± 0%    0.23MB ± 0%  -92.44%  (p=0.008 n=5+5)
FastDelta/testdata/big-heap.pprof-10    35.4MB ± 0%     1.7MB ± 0%  -95.18%  (p=0.008 n=5+5)

name                                  old allocs/op  new allocs/op  delta
FastDelta/testdata/heap.pprof-10         41.2k ± 0%      1.3k ± 0%  -96.74%  (p=0.008 n=5+5)
FastDelta/testdata/big-heap.pprof-10      524k ± 0%       14k ± 0%  -97.25%  (p=0.008 n=5+5)
```

## Change Summary

- vendors [fastdelta](https://github.com/nsrip-dd/fastdelta/)
- fix for `duration_nanos` tracking
- fixes for over-fit test fixtures
- test coverage for the 3 delta implementations (fastdelta, pprofile, comparing)
- comparing telemetry

### CODE REVIEW

**Review by nsrip-dd — CHANGES_REQUESTED** [review]
Thanks Paul!

* There is one breaking change around the `StatsdClient` interface that definitely needs to be addressed. I'm wondering if we should keep the timing stuff, or just the old/new consistency check (so we can still run the implementations side-by-side just to make sure they agree content-wise)
* I've also left some comments more related to the way the code is structured. If those suggestions are too much for this PR, especially related to the nesting of the different pass functions, we can address them in a subsequent PR.

File: profiler/profile_test.go
Comment (pmbauer) [reply profiler/profile_test.go:114]: I'm going to push back on this one.  I can clean this up if we want, but this test caught issues and exercises the deltaProfiler implementation and env var switching.

File: profiler/internal/fastdelta/pprof_records.go
Comment (pmbauer) [reply profiler/internal/fastdelta/pprof_records.go:69]: I'd like to leave them for completeness.  It's unlikely the protobuf def will change.

File: profiler/profile.go
Comment (pmbauer) [reply profiler/profile.go:462]: will do

File: profiler/profiler.go
Comment (pmbauer) [reply profiler/profiler.go:468]: Good point, will fix.
I will leave the change to the comment on Timing though.  The old comment was just wrong.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:154]: Will think about this.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:244]: I'd like to keep the switch/type/case pattern in all spots that parse messages for consistency.  Having this everywhere makes the code more readable and more extensible when we inevitably need to add more cases.

File: internal/env.go
Comment (pmbauer) [reply internal/env.go:47]: It is needed for this, otherwise the `defaultConfig` construction gets more ugly:

https://github.com/DataDog/dd-trace-go/pull/1511/files#diff-8f9f977d509d59fb878b9f086454a6b34dda3d6577d8a4568230af5b9dd8deb3R212

os.Getenv returns both a string and an err.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:116]: one, it's over-fitting for the benchmarks, obv not a good reason. :)  
Removing the sizes as it makes no difference unless we only take one profile per process then quit.

File: internal/env.go
Comment (nsrip-dd) [reply internal/env.go:47]: `os.Getenv` just returns a string ([reference](https://pkg.go.dev/os#Getenv)), are you thinking of `os.LookupEnv`?

Comment (pmbauer) [reply internal/env.go:47]: ah yes, will fix

File: profiler/internal/fastdelta/pprof_records.go
Comment (nsrip-dd) [reply profiler/internal/fastdelta/pprof_records.go:13]: Per our discussion, I'll consider these resolved. We can change this later if we want to, it's not the highest priority for this set of changes.

File: profiler/profile_test.go
Comment (nsrip-dd) [reply profiler/profile_test.go:114]: To capture our conversation for reference, I've suggested that we make a similar change that adds the extra coverage, but keep the diff smaller like this:

```go
func TestRunProfile(t *testing.T) {
    old := os.Getenv("DD_PROFILER_DELTA_IMPL")
    defer os.Setenv("DD_PROFILER_DELTA_IMPL", old)
    for _, deltaMethod := range []string{"fastdelta", "comparing", "pprofile"} {
        os.Setenv("DD_PROFILER_DELTA_IMPL", deltaMethod)
        testRunProfile(t)
    }
}

func testRunProfile(t *testing.T) {
   ... the original function (prior to this PR) ...
}
```

This avoids adding yet another layer of nesting to this test.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:244]: fixed

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:154]: fixed

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:191]: I re-worked this a bit so the functions returning molecule.MessageEachFn only close over the state and pass off to a helper.  This makes the state used in each pass explicit and isolated with one level less of nesting.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:232]: fixed

**Comment by pmbauer** [comment]
@nsrip-dd I've addressed each comment with isolated commits with the message `review: ...` for easy comparison.  I have one more fix to make for https://github.com/DataDog/dd-trace-go/pull/1511/files#diff-2342fcb89653c2a2af7d9fda4d2e9aa2a96dacf2df3a557c8b7b18cb4a499b55R352

**Comment by pmbauer** [comment]
I broke a test, fixing

File: profiler/internal/fastdelta/pprof_records.go
Comment (pmbauer) [reply profiler/internal/fastdelta/pprof_records.go:13]: After writing these types, I noticed [molecule](https://github.com/richardartoul/molecule/blob/master/src/codec/constants.go) follows a similar pattern.

**Comment by nsrip-dd** [comment]
Okay, I think all the stuff from my first review has been addressed. Thanks Paul!

At this point, we should have Felix take a look. We'll mainly want to focus on the following things in my opinion:

* Are we satisfied with the correctness of the algorithm and its implementation?
* Are the tests we have now sufficient? We've tested this on real apps with no fidelity problems, and the performance looks really good. But we can't sustainably do several release cycles with intensive fidelity checking every time we want to further change this. I think we can articulate the risks of this approach and probably come up with a reasonable way to test against them.
  * Off the top of my head, I think the biggest risk is the input dramatically changing. For example, I hard-coded the array that holds the sample values to 4 entries, so we the sample map doesn't have to pointer-chase to dynamically-allocated slices of values. 4 values is the most that a Go profile currently has (the heap profile). What if Go adds more values to a profile (like allocation lifetimes)? We also want to worry about being too trusting of indices in the profile, and reading out of bounds. 

Any style stuff or aesthetic refactoring can wait until a subsequent PR IMO. We should at least be happy with the overall design :)

**Comment by pmbauer** [comment]
@nsrip-dd your review still says 'changes requested'  Is there anything else left to go in before getting a review from @felixge?

**Review by nsrip-dd — DISMISSED** [review]

**Comment by nsrip-dd** [comment]
Nope, all good for me! IMO these changes are OK to merge as-is especially since this is off-by-default behind an environment variable. I'll check in with Felix tomorrow.

**Review by felixge — COMMENTED** [review]
I was hoping to completely review this today, but it's a lot of code 😅. That being said, I made some good progress. Most of my suggestions can be applied by merging this PR: https://github.com/DataDog/dd-trace-go/pull/1550

File: profiler/options.go
Comment (pmbauer) [reply profiler/options.go:214]: In Paris, we agreed to release this behind a flag, disabled by default.  I still think we should release this in stages.

- 1.44.0 with both implementations, old algorithm as default, new algorithm enabled with an env var. This is so we can continue testing internally.  dd-go apps are not going to go to prod with a non-release dd-trace-go.  Customers that need it can opt-in.
- 1.4?.0 still include both algorithms, make fastdelta the default, enable old algorithm with an env var fallback.

File: profiler/profile.go
Comment (pmbauer) [reply profiler/profile.go:384]: This was something I asked for when fastdelta was its own library.
Compression is an orthogonal concern to the process of computing delta's.
You are right `pprofile.ParseData()` dealt with it.  Keeping that out of DeltaComputer is exactly because it is not at the same level of abstraction as `pprofile.ParseData()` and `deltaProf.Write()`.

File: profiler/internal/fastdelta/location_index.go
Comment (pmbauer) [reply profiler/internal/fastdelta/location_index.go:34]: it used to live in `fd.go` and I don't remember why it was moved out; @nsrip-dd, do you remember? I don't think we'd ever expose this to outside callers.

File: profiler/internal/fastdelta/location_index.go
Comment (pmbauer) [reply profiler/internal/fastdelta/location_index.go:20]: Thanks for the fresh take on this. I like it much better.

File: profiler/internal/fastdelta/location_index.go
Comment (nsrip-dd) [reply profiler/internal/fastdelta/location_index.go:34]: I moved it to another file because I felt it was a somewhat trickier component that I'd want to look at on its own. As for exposing it to outside callers, Felix's point is that having the methods we actually intend to use capitalized is a reminder for us of what the intended interface is for this type. If this _did_ live in another package, Insert, Get, and GetMeta would be the methods we actually expose. The rest would be implementation details. That's also reflected in the current tests, which only access `locationIndex` through those methods.

Comment (pmbauer) [reply profiler/internal/fastdelta/location_index.go:34]: SGTM; thanks

File: profiler/profile.go
Comment (felixge) [reply profiler/profile.go:384]: Ok, I don't feel strongly about it.

File: profiler/internal/fastdelta/fd.go
Comment (felixge) [inline profiler/internal/fastdelta/fd.go:697]: 🐞 `dc.hashes` contains hashes of labels and locations. By sorting it, samples with the same set of location ids ordered in different orders will produce the same final hash. I think that's wrong.

**Review by felixge — CHANGES_REQUESTED** [review]
Might have found a bug, PTAL.

File: profiler/internal/fastdelta/fd.go
Comment (nsrip-dd) [reply profiler/internal/fastdelta/fd.go:697]: That makes sense to me. My intention for sorting was to consistently order the _labels_, but you're right that the call stack shouldn't be touched. I'll push a fix

Comment (nsrip-dd) [reply profiler/internal/fastdelta/fd.go:697]: See https://github.com/DataDog/dd-trace-go/pull/1558

File: profiler/internal/fastdelta/fd.go
Comment (felixge) [inline profiler/internal/fastdelta/fd.go:321]: 🐞 Since there is no `includeStringIndexFields` call here, samples with labels that go through this code pass may have string table references that don't exist.

**Review by felixge — CHANGES_REQUESTED** [review]
🐞 Maybe I'm missing something, but it seems like the entire stringsTable implementation assumes that string table ID references remain stable between profiles? I can see how that might be a reasonable assumption for the profiles currently emitted by the go runtime, but I'm worried that it will be very difficult to adept our implementation if this invariant gets broken in the future.

File: profiler/internal/fastdelta/fd.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd.go:321]: good catch, will fix

**Comment by pmbauer** [comment]
> 🐞 Maybe I'm missing something, but it seems like the entire stringsTable implementation assumes that string table ID references remain stable between profiles?

The stringTable is not used between profiles and is [reset](https://github.com/DataDog/dd-trace-go/blob/e8bfed74c149c818476da7d9644096e771c1d341/profiler/internal/fastdelta/fd.go#L143) at the start of each delta.  Unless I'm misunderstanding something, we don't assume the IDs are stable between profiles.

File: profiler/internal/fastdelta/fd_test.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd_test.go:55]: I agree. We could re-work this and even incorporate `b.N` into the test input.  I don't think that would be too difficult.

> It doesn't measure any delta computation.

Do you mean the [one actual delta in the bench](https://github.com/DataDog/dd-trace-go/blob/1da2b74838c81b4038a5fe64cb3d6a92a778361e/profiler/internal/fastdelta/fd_test.go#L52) results in no non-zero samples for the [value types](https://github.com/DataDog/dd-trace-go/blob/1da2b74838c81b4038a5fe64cb3d6a92a778361e/profiler/internal/fastdelta/fd_test.go#L45-L46) configured?  The table you have is only showing the set of non-`alloc_*` samples.

Comment (pmbauer) [reply profiler/internal/fastdelta/fd_test.go:55]: As there is only one `Delta` computation, the current benchmark is under-representing the gains for processes that last longer than 2 profiling cycles.

Comment (felixge) [reply profiler/internal/fastdelta/fd_test.go:55]: > As there is only one `Delta` computation, the current benchmark is under-representing the gains for processes that last longer than 2 profiling cycles.

Yes, that's another problem with this benchmark! I think it'd be better if it create a single DeltaComputer and then applied pprofs against it in the loop rather than creating a new computer in each loop.

**Comment by felixge** [comment]
>  The strings that get written out to the delta pprof will have the same offsets as the input pprof passed to Delta, but there is no assumed carry-over from one delta pprof to the next.

💡 Ah, got it. For some reason I thought that there might be references into the string table from the previous profile. But I can see now that this is clearly not the case. All references are from the current profile, so it doesn't matter if the same string had a different index in the previous profile. ✅

File: profiler/internal/fastdelta/fd.go
Comment (felixge) [reply profiler/internal/fastdelta/fd.go:321]: > good catch
> In practice, we throw away the first delta and the comparingDeltaProfiler didn't catch this one, but will fix.

Maybe hold off on that fix. I have a branch with some more refactorings where I can easily fix this as well.

File: profiler/internal/fastdelta/fd_test.go
Comment (pmbauer) [reply profiler/internal/fastdelta/fd_test.go:55]: Agreed.  This bench could use some more work.  I'll have some updates later today.

**Review by nsrip-dd — APPROVED** [review]

**Review by felixge — APPROVED** [review]

**Comment by zdyj3170101136** [comment]
The delta profile I designed relies on the built-in ordering of go performance profiles, which does not consume cpu.

please see https://github.com/DataDog/dd-trace-go/issues/2804.

**Pull request merged**
